### PR TITLE
Change extensions npiid

### LINKED ISSUE #772: There should be a fixed NPI ID for Program Info and Attach/Hook NPI.
Currently a different NPI ID is used for every program type and attach type. Since this GUID is provided by untrusted UM app to EC, the EC may inadvertently try to attach with a NPI provider unrelated to eBPF.

This implementation should be changed such that the same GUID as NPI ID for program info and attach/hook NPI.

The NPI Providers should pass the program ID or attach ID in NPI_REGISTRATION_INSTANCE structure. EC should check for matching program ID or attach ID in the NPI_CLIENT_ATTACH_PROVIDER_FN callback before calling NmrClientAttachProvider. As an optimization the provider module ID can be set to the program ID/attach ID.

### PR DESCRIPTION
**Pull request by shankarseal**
## Description

Fixes #772 .
New fixed IIDs are defined for program information and hook extensions. Providers pass program type and attach type as module IDs.

## Testing

A new api_test was added to load 2 programs of different types together. Existing tests also provided coverage.

## Documentation

ebpfExtensions.md is updated in this PR.

### CODE REVIEW

**Review by dthaler — COMMENTED** [review]
don't delete dependabot.yml in this PR.
You can delete it in your fork's main branch but should not be deleted in any PR.

**Comment by shankarseal** [comment]
> don't delete dependabot.yml in this PR. You can delete it in your fork's main branch but should not be deleted in any PR.

That was unintentional. FIxed.

File: netebpfext/net_ebpf_ext.h [hunk 1]
```diff
@@ -36,6 +36,7 @@
 #include <netiodef.h>
 #include <ntddk.h>
 
+#include "ebpf_extension_uuids.h"
```
[the remark this answers is not in the document]
Comment (shankarseal) [reply netebpfext/net_ebpf_ext.h]: Moved to C files that need this.

File: tests/sample/ext/drv/sample_ext.c [hunk 2]
```diff
@@ -298,6 +296,7 @@ sample_ebpf_extension_program_info_provider_register()
                          .ProviderRegistrationInstance.NpiSpecificCharacteristics;
     program_data = (ebpf_program_data_t*)extension_data->data;
     program_data->program_info->program_type_descriptor.program_type = EBPF_PROGRAM_TYPE_SAMPLE;
+    _sample_ebpf_extension_program_info_provider_moduleid.Guid = EBPF_PROGRAM_TYPE_SAMPLE;
```
[the remark this answers is not in the document]
Comment (shankarseal) [reply tests/sample/ext/drv/sample_ext.c:299]: I keep getting C2099. Tried a lot of things. Nothing works. I think we have to change the constant variables to `#define` to make C compiler happy. See: https://stackoverflow.com/questions/3025050/error-initializer-element-is-not-constant-when-trying-to-initialize-variable-w

File: netebpfext/net_ebpf_ext_hook_provider.c [hunk 3]
```diff
@@ -3,6 +3,7 @@
 
 #include <wdm.h>
 #include "net_ebpf_ext_hook_provider.h"
+#include "ebpf_extension_uuids.h"
```
Comment (dthaler) [inline netebpfext/net_ebpf_ext_hook_provider.c:6]: move up one line so that the `""` includes are in alphabetical order

**Review by dthaler — APPROVED** [review]

File: scripts/config_test_vm.psm1 [hunk 4]
```diff
@@ -220,24 +220,16 @@ function Import-ResultsFromVM
...
-        }
-
         Invoke-Command -Session $VMSession -ScriptBlock {
             if (!(Test-Path "$Env:SystemDrive\KernelDumps")) {
                 New-Item -ItemType Directory -Path "$Env:SystemDrive\KernelDumps"
             }
             Move-Item $Env:WinDir\*.dmp $Env:SystemDrive\KernelDumps -ErrorAction Ignore
         }
-        Copy-Item -FromSession $VMSession "$VMSystemDrive\KernelDumps" -Destination ".\TestLogs\$VMName\KernelDumps" -Force -ErrorAction Ignore 2>&1 | Write-Log
+        Copy-Item -FromSession $VMSession "$VMSystemDrive\KernelDumps" -Destination ".\TestLogs\$VMName" -Recurse -Force -ErrorAction Ignore 2>&1 | Write-Log
```
Comment (Alan-Jowett) [inline scripts/config_test_vm.psm1:229]: How is this related to changing the module_id? If it's not, can we move it to a different PR.

**Review by Alan-Jowett — APPROVED** [review]
Looks good, still curious about the changes in config_test_vm.psm1 as they seem unrelated.

File: libs/platform/ebpf_platform.h [hunk 5]
```diff
@@ -736,7 +736,8 @@ extern "C"
      * @brief Load an extension and get its dispatch table.
      *
      * @param[out] client_context Context used to unload the extension.
-     * @param[in] interface_id GUID representing the identity of the interface.
+     * @param[in] interface_id GUID representing the identity of the extension interface.
+     * @param[in] expected_provider_module_id GUID representing the expected identity of the provider.
```
[the remark this answers is not in the document]
Comment (shankarseal) [reply libs/platform/ebpf_platform.h:740]: OK - I will think of a better adjective and replace in a future PR. Its not an input to the NMR handshake process. So I dont think "requested" is right either.
